ydb/core/tx/replication/service/s3_writer.cpp

```diff
@@ -0,0 +1,239 @@
+        if (CanRetry(error)) {
+            Retry();
+        } else {
+            Leave(TStringBuilder() << "S3 error: " << error.GetMessage().c_str());
+        }
+    }
+
+    void Leave(const TString& error) {
+        CB_LOG_I("Leave"
+            << ": error# " << error);
+
+        // TODO support different error kinds
+        Send(Worker, new TEvWorker::TEvGone(TEvWorker::TEvGone::S3_ERROR));
+
+        PassAway();
+    }
+
+    void SendS3Request() {
+        Y_VERIFY(RequestInFlight);
+        Send(S3Client, new TEvExternalStorage::TEvPutObjectRequest(RequestInFlight->Request, TString(RequestInFlight->Buffer)));
+    }
+
+    void Handle(TEvWorker::TEvHandshake::TPtr& ev) {
+        Worker = ev->Sender;
+        CB_LOG_D("Handshake"
+            << ": worker# " << Worker);
+
+        S3Client = RegisterWithSameMailbox(NWrappers::CreateS3Wrapper(ExternalStorageConfig->ConstructStorageOperator()));
+
+        WriteIdentity();
+    }
+
+    void WriteIdentity() {
+        TString key = Sprintf("writer.%s.json", WriterName.c_str());
+
+        auto request = Aws::S3::Model::PutObjectRequest()
+            .WithKey(key);
+
+        TString buffer = Sprintf(R"({"table_name":"%s","writer_name":"%s","finished":%s})", TableName.c_str(), WriterName.c_str(), Finished ? "true" : "false");
+
+        RequestInFlight = std::make_unique<TS3Request>(request, buffer);
+        SendS3Request();
+    }
+
+    void PassAway() override {
+        Send(S3Client, new TEvents::TEvPoison());
+        TActor::PassAway();
+    }
+
+    void Handle(TEvWorker::TEvData::TPtr& ev) {
+        CB_LOG_D("Handle " << ev->Get()->ToString());
+
+        if (!ev->Get()->Records) {
+            Finished = true;
+            WriteIdentity();
+            return;
+        }
+
+        TString key = Sprintf("part.%ld.%s.jsonl", ev->Get()->Records[0].Offset, WriterName.c_str());
+
+        auto request = Aws::S3::Model::PutObjectRequest()
+            .WithKey(key);
+
+        TVector<TString> data;
+
+        for (auto& rec : ev->Get()->Records) {
+            data.push_back(rec.Data);
+        }
+
+        TString buffer = JoinSeq('\n', data) + "\n";
+
+        RequestInFlight = std::make_unique<TS3Request>(request, buffer);
+        SendS3Request();
+    }
+
+    void Handle(TEvExternalStorage::TEvPutObjectResponse::TPtr& ev) {
+        const auto& result = ev->Get()->Result;
+
+        CB_LOG_D("HandleScheme TEvExternalStorage::TEvPutObjectResponse"
+            << ", result# " << result);
+
+        if (!CheckResult(result, TStringBuf("PutObject"))) {
+            return;
+        } else {
+            RequestInFlight = nullptr;
+        }
+
+        if (IdentityWritten && !Finished) {
+            Send(Worker, new TEvWorker::TEvPoll());
+        } else if (!Finished) {
+            IdentityWritten = true;
+            Send(Worker, new TEvWorker::TEvHandshake());
+        } else {
+            Send(Worker, new TEvWorker::TEvGone(TEvWorker::TEvGone::DONE));
+        }
```

Comment:
Вот так было бы понятнее:
```suggestion
        if (!IdentityWritten) {
            IdentityWritten = true;
            Send(Worker, new TEvWorker::TEvHandshake());
        } else if (!Finished) {
            Send(Worker, new TEvWorker::TEvPoll());
        } else {
            Send(Worker, new TEvWorker::TEvGone(TEvWorker::TEvGone::DONE));
        }
```